pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -664,8 +664,16 @@ func findPortFromMetadata(svcPort v1.ServicePort, podPorts []model.PodPort) (int
 			}
 		}
 	case intstr.Int:
-		// For a direct reference we can just return the port number
-		return target.IntValue(), nil
+		num := int(target.IntVal)
+		if len(podPorts) > 0 {
+			for _, port := range podPorts {
+				if port.ContainerPort == num {
```

Comment:
I don't really get what this is supposed to do. container port is an optional field. Why would we ever not create a listener for a service?

Reply:
This is inbound listener. We should not create inbound listener for port that service is not listening on/does not have a target port on. We do that validation and skip for named ports but see string case above, but if that target port is int, we just take it as is and create service instance and there by inbound listener.

Reply:
That is because you cannot know if the pod is listening on the port or not. It is impossible. `containerPort` is an optional field, it is NOT required for a pod to add this

Reply:
See the example config_dump for a service that it creates listener on port `7443` (because it has a container port `7443` and service instance for it is created using `FindPort`(which looks at container ports) here https://github.com/istio/istio/blob/master/pilot/pkg/serviceregistry/kube/controller/controller.go#L693 when pod is available and if it is not available we get it from metadata here https://github.com/istio/istio/blob/master/pilot/pkg/serviceregistry/kube/controller/controller.go#L629 (Which is again container ports)

```
   {
     "name": "virtualInbound",
     "active_state": {
      "version_info": "2020-01-13T14:45:18Z/2935",
      "listener": {
       "name": "virtualInbound",
       "address": {
        "socket_address": {
         "address": "0.0.0.0",
         "port_value": 15006
        }
       },
       "filter_chains": [
        {
         "filter_chain_match": {
          "prefix_ranges": [
           {
            "address_prefix": "0.0.0.0",
            "prefix_len": 0
           }
          ]
         },
         "filters": [
          {
           "name": "envoy.tcp_proxy",
           "typed_config": {
            "@type": "type.googleapis.com/envoy.config.filter.network.tcp_proxy.v2.TcpProxy",
            "stat_prefix": "InboundPassthroughClusterIpv4",
            "access_log": [
             {
              "name": "envoy.file_access_log",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.config.accesslog.v2.FileAccessLog",
               "path": "/dev/stdout",
               "format": "[%START_TIME%] \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%\" %RESPONSE_CODE% %RESPONSE_FLAGS% \"%DYNAMIC_METADATA(istio.mixer:status)%\" \"%UPSTREAM_TRANSPORT_FAILURE_REASON%\" %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% \"%REQ(X-FORWARDED-FOR)%\" \"%REQ(USER-AGENT)%\" \"%REQ(X-REQUEST-ID)%\" \"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\" %UPSTREAM_CLUSTER% %UPSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_REMOTE_ADDRESS% %REQUESTED_SERVER_NAME% %ROUTE_NAME%\n"
              }
             }
            ],
            "cluster": "InboundPassthroughClusterIpv4"
           }
          }
         ],
         "metadata": {
          "filter_metadata": {
           "pilot_meta": {
            "original_listener_name": "virtualInbound"
           }
          }
         }
        },
        {
         "filter_chain_match": {
          "prefix_ranges": [
           {
            "address_prefix": "::0",
            "prefix_len": 0
           }
          ]
         },
         "filters": [
          {
           "name": "envoy.tcp_proxy",
           "typed_config": {
            "@type": "type.googleapis.com/envoy.config.filter.network.tcp_proxy.v2.TcpProxy",
            "stat_prefix": "InboundPassthroughClusterIpv6",
            "access_log": [
             {
              "name": "envoy.file_access_log",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.config.accesslog.v2.FileAccessLog",
               "path": "/dev/stdout",
               "format": "[%START_TIME%] \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%\" %RESPONSE_CODE% %RESPONSE_FLAGS% \"%DYNAMIC_METADATA(istio.mixer:status)%\" \"%UPSTREAM_TRANSPORT_FAILURE_REASON%\" %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% \"%REQ(X-FORWARDED-FOR)%\" \"%REQ(USER-AGENT)%\" \"%REQ(X-REQUEST-ID)%\" \"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\" %UPSTREAM_CLUSTER% %UPSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_REMOTE_ADDRESS% %REQUESTED_SERVER_NAME% %ROUTE_NAME%\n"
              }
             }
            ],
            "cluster": "InboundPassthroughClusterIpv6"
           }
          }
         ],
         "metadata": {
          "filter_metadata": {
           "pilot_meta": {
            "original_listener_name": "virtualInbound"
           }
          }
         }
        },
        {
         "filter_chain_match": {
          "prefix_ranges": [
           {
            "address_prefix": "10.251.130.234",
            "prefix_len": 32
           }
          ],
          "destination_port": 15020
         },
         "filters": [
          {
           "name": "envoy.tcp_proxy",
           "typed_config": {
            "@type": "type.googleapis.com/envoy.config.filter.network.tcp_proxy.v2.TcpProxy",
            "stat_prefix": "inbound|15020|mgmt-15020|mgmtCluster",
            "access_log": [
             {
              "name": "envoy.file_access_log",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.config.accesslog.v2.FileAccessLog",
               "path": "/dev/stdout",
               "format": "[%START_TIME%] \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%\" %RESPONSE_CODE% %RESPONSE_FLAGS% \"%DYNAMIC_METADATA(istio.mixer:status)%\" \"%UPSTREAM_TRANSPORT_FAILURE_REASON%\" %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% \"%REQ(X-FORWARDED-FOR)%\" \"%REQ(USER-AGENT)%\" \"%REQ(X-REQUEST-ID)%\" \"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\" %UPSTREAM_CLUSTER% %UPSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_REMOTE_ADDRESS% %REQUESTED_SERVER_NAME% %ROUTE_NAME%\n"
              }
             }
            ],
            "cluster": "inbound|15020|mgmt-15020|mgmtCluster"
           }
          }
         ],
         "metadata": {
          "filter_metadata": {
           "pilot_meta": {
            "original_listener_name": "10.251.130.234_15020"
           }
          }
         }
        },
        {
         "filter_chain_match": {
          "prefix_ranges": [
           {
            "address_prefix": "10.251.130.234",
            "prefix_len": 32
           }
          ],
          "destination_port": 7443
         },
         "filters": [
          {
           "name": "envoy.http_connection_manager",
           "typed_config": {
            "@type": "type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager",
            "stat_prefix": "inbound_10.251.130.234_7443",
            "http_filters": [
             {
              "name": "istio_authn",
              "typed_config": {
               "@type": "type.googleapis.com/istio.envoy.config.filter.http.authn.v2alpha1.FilterConfig",
               "policy": {
                "peers": [
                 {
                  "mtls": {}
                 }
                ],
                "peer_is_optional": true
               }
              }
             },
             {
              "name": "envoy.filters.http.grpc_stats",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.config.filter.http.grpc_stats.v2alpha.FilterConfig",
               "emit_filter_state": true
              }
             },
             {
              "name": "envoy.cors"
             },
             {
              "name": "envoy.fault"
             },
             {
              "name": "envoy.router"
             }
            ],
            "tracing": {
             "client_sampling": {
              "value": 100
             },
             "random_sampling": {
              "value": 1
             },
             "overall_sampling": {
              "value": 100
             }
            },
            "http2_protocol_options": {},
            "server_name": "istio-envoy",
            "access_log": [
             {
              "name": "envoy.file_access_log",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.config.accesslog.v2.FileAccessLog",
               "path": "/dev/stdout",
               "format": "[%START_TIME%] \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%\" %RESPONSE_CODE% %RESPONSE_FLAGS% \"%DYNAMIC_METADATA(istio.mixer:status)%\" \"%UPSTREAM_TRANSPORT_FAILURE_REASON%\" %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% \"%REQ(X-FORWARDED-FOR)%\" \"%REQ(USER-AGENT)%\" \"%REQ(X-REQUEST-ID)%\" \"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\" %UPSTREAM_CLUSTER% %UPSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_REMOTE_ADDRESS% %REQUESTED_SERVER_NAME% %ROUTE_NAME%\n"
              }
             }
            ],
            "use_remote_address": false,
            "generate_request_id": true,
            "forward_client_cert_details": "APPEND_FORWARD",
            "set_current_client_cert_details": {
             "subject": true,
             "dns": true,
             "uri": true
            },
            "upgrade_configs": [
             {
              "upgrade_type": "websocket"
             }
            ],
            "stream_idle_timeout": "0s",
            "normalize_path": true,
            "route_config": {
             "name": "inbound|7443|grpc|route-update-service.service-mesh.svc.cluster.local",
             "virtual_hosts": [
              {
               "name": "inbound|http|7443",
               "domains": [
                "*"
               ],
               "routes": [
                {
                 "match": {
                  "prefix": "/"
                 },
                 "decorator": {
                  "operation": "route-update-service.service-mesh.svc.cluster.local:7443/*"
                 },
                 "name": "default",
                 "route": {
                  "timeout": "0s",
                  "max_grpc_timeout": "0s",
                  "cluster": "inbound|7443|grpc|route-update-service.service-mesh.svc.cluster.local"
                 }
                }
               ]
              }
             ],
             "validate_clusters": false
            }
           }
          }
         ],
         "metadata": {
          "filter_metadata": {
           "pilot_meta": {
            "original_listener_name": "10.251.130.234_7443"
           }
          }
         },
         "transport_socket": {
          "name": "envoy.transport_sockets.tls",
          "typed_config": {
           "@type": "type.googleapis.com/envoy.api.v2.auth.DownstreamTlsContext",
           "common_tls_context": {
            "tls_certificates": [
             {
              "certificate_chain": {
               "filename": "/server-certs/server/certificates/server.pem"
              },
              "private_key": {
               "filename": "/server-certs/server/keys/server-key.pem"
              }
             }
            ],
            "alpn_protocols": [
             "h2",
             "http/1.1"
            ],
            "validation_context": {
             "trusted_ca": {
              "filename": "/server-certs/ca.pem"
             }
            }
           },
           "require_client_certificate": true
          }
         }
        },
        {
         "filter_chain_match": {
          "prefix_ranges": [
           {
            "address_prefix": "fe80::42:aff:fefb:82ea",
            "prefix_len": 128
           }
          ],
          "destination_port": 7443
         },
         "filters": [
          {
           "name": "envoy.http_connection_manager",
           "typed_config": {
            "@type": "type.googleapis.com/envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager",
            "stat_prefix": "inbound_fe80::42:aff:fefb:82ea_7443",
            "http_filters": [
             {
              "name": "istio_authn",
              "typed_config": {
               "@type": "type.googleapis.com/istio.envoy.config.filter.http.authn.v2alpha1.FilterConfig",
               "policy": {
                "peers": [
                 {
                  "mtls": {}
                 }
                ],
                "peer_is_optional": true
               }
              }
             },
             {
              "name": "envoy.filters.http.grpc_stats",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.config.filter.http.grpc_stats.v2alpha.FilterConfig",
               "emit_filter_state": true
              }
             },
             {
              "name": "envoy.cors"
             },
             {
              "name": "envoy.fault"
             },
             {
              "name": "envoy.router"
             }
            ],
            "tracing": {
             "client_sampling": {
              "value": 100
             },
             "random_sampling": {
              "value": 1
             },
             "overall_sampling": {
              "value": 100
             }
            },
            "http2_protocol_options": {},
            "server_name": "istio-envoy",
            "access_log": [
             {
              "name": "envoy.file_access_log",
              "typed_config": {
               "@type": "type.googleapis.com/envoy.config.accesslog.v2.FileAccessLog",
               "path": "/dev/stdout",
               "format": "[%START_TIME%] \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)% %PROTOCOL%\" %RESPONSE_CODE% %RESPONSE_FLAGS% \"%DYNAMIC_METADATA(istio.mixer:status)%\" \"%UPSTREAM_TRANSPORT_FAILURE_REASON%\" %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)% \"%REQ(X-FORWARDED-FOR)%\" \"%REQ(USER-AGENT)%\" \"%REQ(X-REQUEST-ID)%\" \"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\" %UPSTREAM_CLUSTER% %UPSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_REMOTE_ADDRESS% %REQUESTED_SERVER_NAME% %ROUTE_NAME%\n"
              }
             }
            ],
            "use_remote_address": false,
            "generate_request_id": true,
            "forward_client_cert_details": "APPEND_FORWARD",
            "set_current_client_cert_details": {
             "subject": true,
             "dns": true,
             "uri": true
            },
            "upgrade_configs": [
             {
              "upgrade_type": "websocket"
             }
            ],
            "stream_idle_timeout": "0s",
            "normalize_path": true,
            "route_config": {
             "name": "inbound|7443|grpc|route-update-service.service-mesh.svc.cluster.local",
             "virtual_hosts": [
              {
               "name": "inbound|http|7443",
               "domains": [
                "*"
               ],
               "routes": [
                {
                 "match": {
                  "prefix": "/"
                 },
                 "decorator": {
                  "operation": "route-update-service.service-mesh.svc.cluster.local:7443/*"
                 },
                 "name": "default",
                 "route": {
                  "timeout": "0s",
                  "max_grpc_timeout": "0s",
                  "cluster": "inbound|7443|grpc|route-update-service.service-mesh.svc.cluster.local"
                 }
                }
               ]
              }
             ],
             "validate_clusters": false
            }
           }
          }
         ],
         "metadata": {
          "filter_metadata": {
           "pilot_meta": {
            "original_listener_name": "fe80::42:aff:fefb:82ea_7443"
           }
          }
         },
         "transport_socket": {
          "name": "envoy.transport_sockets.tls",
          "typed_config": {
           "@type": "type.googleapis.com/envoy.api.v2.auth.DownstreamTlsContext",
           "common_tls_context": {
            "tls_certificates": [
             {
              "certificate_chain": {
               "filename": "/server-certs/server/certificates/server.pem"
              },
              "private_key": {
               "filename": "/server-certs/server/keys/server-key.pem"
              }
             }
            ],
            "alpn_protocols": [
             "h2",
             "http/1.1"
            ],
            "validation_context": {
             "trusted_ca": {
              "filename": "/server-certs/ca.pem"
             }
            }
           },
           "require_client_certificate": true
          }
         }
        }
       ],
       "listener_filters": [
        {
         "name": "envoy.listener.original_dst"
        },
        {
         "name": "envoy.listener.tls_inspector"
        }
       ],
       "listener_filters_timeout": "1s",
       "traffic_direction": "INBOUND",
       "continue_on_listener_filters_timeout": true
      },
      "last_updated": "2020-01-13T14:46:19.071Z"
     }
    }
   ]
  },
```

I agree  they are optional - but looks if they are available we do pickup and create listeners for them. 

And for named ports, if they are not available we skip - Should we do the same for int ports?  Is it not inconsistent?